tests/test_linter.py

```diff
@@ -1,8 +1,340 @@
 
 
-def test_check(data, txvars):
-    c = Check(data, txvars)
+def test_parser():
+    t = 'SecRule REQUEST_HEADERS:User-Agent "@rx ^Mozilla" "id:1,phase:1,log,status:403"'
+    p = parse_config(t)
+
+    assert p is not None
+
+
+def test_check_ignore_proper_case():
+    t = 'SecRule REQUEST_HEADERS:User-Agent "@rx ^Mozilla" "id:1,phase:1,log,status:403"'
+    p = parse_config(t)
+    c = Check(p)
     c.check_ignore_case()
 
     assert len(c.caseerror) == 0
+
+
+def test_check_ignore_case_fail_invalid_action_case():
+    """ Two actions are in the wrong case. """
+    t = 'SecRule REQUEST_HEADERS:User-Agent "@rx ^Mozilla" "id:1,phase:1,LOG,NoLOg,status:403"'
+    p = parse_config(t)
+    c = Check(p)
+    c.check_ignore_case()
+
+    assert len(c.caseerror) == 2
+
+
+def test_check_action_order():
+    """ Test that the actions are in the correct order. """
+    t = 'SecRule REQUEST_HEADERS:User-Agent "@rx ^Mozilla" "id:1,phase:1,nolog"'
+    p = parse_config(t)
+    c = Check(p)
+    c.check_action_order()
+
+    assert len(c.orderacts) == 0
+
+
+def test_check_action_fail_wrong_order():
+    """ Test if the action is in the wrong order. status should go before log """
+    t = 'SecRule REQUEST_HEADERS:User-Agent "@rx ^Mozilla" "id:1,phase:1,log,status:403"'
+    p = parse_config(t)
+    c = Check(p)
+    c.check_action_order()
+
+    assert len(c.orderacts) == 1
+
+
+def test_check_ctl_auditctl_log_parts():
+    """ Test that there is no ctl:auditLogParts action in any rules"""
+    t = 'SecRule REQUEST_HEADERS:User-Agent "@rx ^Mozilla" "id:1,phase:1,log,status:403"'
+    p = parse_config(t)
+    c = Check(p)
+    c.check_ctl_audit_log()
+
+    assert len(c.auditlogparts) == 0
+
+
+def test_check_wrong_ctl_audit_log_parts():
+    t = 'SecRule REQUEST_HEADERS:User-Agent "@rx ^Pizza" "id:1,phase:1,log,ctl:auditLogParts=+E"'
+    p = parse_config(t)
+    c = Check(p)
+    c.check_ctl_audit_log()
+
+    assert len(c.auditlogparts) == 1
+
+
+def test_check_tx_variable():
+    """ Test that variables are defined in the transaction """
+    t = """SecRule &TX:blocking_paranoia_level "@eq 0" \
+    "id:901120,\
+    phase:1,\
+    pass,\
+    nolog,\
+    ver:'OWASP_CRS/4.0.0-rc1',\
+    setvar:'tx.blocking_paranoia_level=1'"
+
+SecRule &TX:detection_paranoia_level "@eq 0" \
+    "id:901125,\
+    phase:1,\
+    pass,\
+    nolog,\
+    ver:'OWASP_CRS/4.0.0-rc1',\
+    setvar:'tx.detection_paranoia_level=%{TX.blocking_paranoia_level}'"
+    """
+    p = parse_config(t)
+    c = Check(p)
+    c.check_tx_variable()
+
+    assert len(c.undef_txvars) == 0
+
+
+def test_check_tx_variable_fail_nonexisting():
+    t = """SecRule TX:foo "@rx bar" \
+    "id:1001,\
+    phase:1,\
+    pass,\
+    nolog"
+
+SecRule ARGS "@rx ^.*$" \
+    "id:1002,\
+    phase:1,\
+    pass,\
+    nolog,\
+    setvar:tx.bar=1"
+        """
+    p = parse_config(t)
+    c = Check(p)
+    c.collect_tx_variable()
+    c.check_tx_variable()
+
+    # Shouldn't this be 2?
```

Comment:
@airween From https://github.com/coreruleset/crs-linter/blob/main/README.md?plain=1#L401-L421, I feel this should be 2 instead of 1?

Reply:
I don't think so. Below in next line you check that the length of `c.undef_txvars` is 1, and that's the correct value, because there is no other `tx` variable were set.